core/reorder/mc64.cpp

```diff
@@ -0,0 +1,148 @@
+#include <ginkgo/core/matrix/sparsity_csr.hpp>
+
+
+#include "core/matrix/csr_kernels.hpp"
+#include "core/reorder/mc64_kernels.hpp"
+
+
+namespace gko {
+namespace reorder {
+namespace mc64 {
+namespace {
+
+
+GKO_REGISTER_OPERATION(initialize_weights, mc64::initialize_weights);
+GKO_REGISTER_OPERATION(initial_matching, mc64::initial_matching);
+GKO_REGISTER_OPERATION(shortest_augmenting_path,
+                       mc64::shortest_augmenting_path);
+GKO_REGISTER_OPERATION(compute_scaling, mc64::compute_scaling);
+
+
+}  // anonymous namespace
+}  // namespace mc64
+
+
+template <typename ValueType, typename IndexType>
+void Mc64<ValueType, IndexType>::generate(std::shared_ptr<const Executor>& exec,
+                                          std::shared_ptr<LinOp> system_matrix)
+{
+    auto mtx = as<matrix_type>(system_matrix);
+    size_type num_rows = mtx->get_size()[0];
+    size_type nnz = mtx->get_num_stored_elements();
+
+    // A real valued array with space for:
+    //     - nnz entries for weights
+    //     - num_rows entries each for the dual vector u, distance information
+    //       and the max weight per row
+    array<remove_complex<ValueType>> value_workspace{exec, nnz + 3 * num_rows};
+    // A zero initialized index array with space for n entries each for parent
+    // information, priority queue handles, generation information, marked
+    // columns, indices corresponding to matched columns in the according row
+    // and still unmatched rows
+    array<IndexType> index_workspace{exec, 6 * num_rows};
+    index_workspace.fill(0);
+
+    array<IndexType> permutation{exec, num_rows};
+    array<IndexType> inv_permutation{exec, num_rows};
+    permutation.fill(-one<IndexType>());
+    inv_permutation.fill(-one<IndexType>());
+
+    const auto row_ptrs = mtx->get_const_row_ptrs();
+    const auto col_idxs = mtx->get_const_col_idxs();
+
+    exec->run(mc64::make_initialize_weights(mtx.get(), value_workspace,
+                                            parameters_.strategy));
+
+    // Compute an initial extreme matching from the nonzero entries for which
+    // the reduced weight (W(i, j) - u(j) - v(i)) is zero. Here, W is the
+    // weight matrix and u and v are the dual vectors. Note that v initially
+    // only contains zeros and hence can still be ignored here.
+    exec->run(mc64::make_initial_matching(
+        num_rows, row_ptrs, col_idxs, value_workspace, permutation,
+        inv_permutation, index_workspace, parameters_.tolerance));
+
+    // For each row that is not contained in the initial matching, search for
+    // an augmenting path, update the matching and compute the new entries
+    // of the dual vectors.
+    addressable_priority_queue<remove_complex<ValueType>, IndexType> Q{
+        parameters_.log2_degree};
+    std::vector<IndexType> q_j{};
+    const auto unmatched = index_workspace.get_data() + 5 * num_rows;
+    auto um = 0;
+    auto root = unmatched[um];
```

Comment:
Would it make sense to store the unmatched rows in a consecutive vector since it should be much smaller than num_rows after the initial matching. Within the `initial_matching `a temporary array could be used. I guess performance wise the difference would be negligible since `shortest_augmenting_path` would be the dominant factor.

Reply:
I would prefer leaving the current version even if it needs more memory. When implementing, I found this being the fastest alternative on the systems I tested. `shortest_augmenting_path` is the dominant factor for a lot of matrices, but definitely not for all, especially when the initial matching already matches (almost) all nodes.